daemon/api_download.go

```diff
@@ -102,55 +108,154 @@ func postSnapDownload(c *Command, r *http.Request, user *auth.UserState) Respons
 }
 
 func streamOneSnap(c *Command, action snapDownloadAction, user *auth.UserState) Response {
-	theStore := getStore(c)
-	var info *snap.Info
-
-	// XXX: once we're HMAC'ing, we only do this bit on the first pass
-	actions := []*store.SnapAction{{
-		Action:       "download",
-		InstanceName: action.SnapName,
-		Revision:     action.Revision,
-		CohortKey:    action.CohortKey,
-		Channel:      action.Channel,
-	}}
-	sars, err := theStore.SnapAction(context.TODO(), nil, actions, user, nil)
+	secret, err := downloadTokensSecret(c)
 	if err != nil {
-		return errToResponse(err, []string{action.SnapName}, InternalError, "cannot download snap: %v")
-	}
-	if len(sars) != 1 {
-		return InternalError("internal error: unexpected number %v of results for a single download", len(sars))
+		return InternalError(err.Error())
 	}
-	info = sars[0].Info
-	downloadInfo := info.DownloadInfo
+	theStore := getStore(c)
 
-	// XXX: this bit goes away once we HMAC
-	if action.ResumeStamp != "" && downloadInfo.Sha3_384 != action.ResumeStamp {
-		return BadRequest("snap to download has different hash")
+	var ss *snapStream
+	if action.ResumeToken == "" {
+		var info *snap.Info
+		actions := []*store.SnapAction{{
+			Action:       "download",
+			InstanceName: action.SnapName,
+			Revision:     action.Revision,
+			CohortKey:    action.CohortKey,
+			Channel:      action.Channel,
+		}}
+		sars, err := theStore.SnapAction(context.TODO(), nil, actions, user, nil)
+		if err != nil {
+			return errToResponse(err, []string{action.SnapName}, InternalError, "cannot download snap: %v")
+		}
+		if len(sars) != 1 {
+			return InternalError("internal error: unexpected number %v of results for a single download", len(sars))
+		}
+		info = sars[0].Info
+
+		ss, err = newSnapStream(action.SnapName, info, secret)
+		if err != nil {
+			return InternalError(err.Error())
+		}
+	} else {
+		var err error
+		ss, err = newResumingSnapStream(action.SnapName, action.ResumeToken, secret)
+		if err != nil {
+			return BadRequest(err.Error())
+		}
+		ss.resume = action.resumePosition
 	}
 
-	rsp := newSnapStream(action.SnapName, info, action.resumePosition)
 	if !action.HeaderPeek {
-		r, s, err := theStore.DownloadStream(context.TODO(), action.SnapName, &downloadInfo, action.resumePosition, user)
+		r, s, err := theStore.DownloadStream(context.TODO(), action.SnapName, ss.Info, action.resumePosition, user)
 		if err != nil {
 			return InternalError(err.Error())
 		}
-		rsp.stream = r
+		ss.stream = r
 		if s != 206 {
 			// store/cdn has no partial content (valid
 			// reply per RFC)
 			logger.Debugf("store refused our range request")
-			rsp.resume = 0
+			ss.resume = 0
 		}
 	}
 
-	return rsp
+	return ss
+}
+
+func newSnapStream(snapName string, info *snap.Info, secret []byte) (*snapStream, error) {
+	dlInfo := &info.DownloadInfo
```

Comment:
One (potentially silly) question. Could we run into a race here? i.e. we acquire the download url, hand it to the client, the client does it thing and in the meantime in the store this snap that we try to download is unpublished for some reason. Would the store still allow us to download the url when we resume in newResumingSnapStream?

Reply:
generally yes, the rules are not fully specified, but yes because of cohorts and epoch support etc